packages/nimble-components/src/rich-text/specs/mention-hld.md

```diff
@@ -0,0 +1,721 @@
+
+The `mention-update` event will also be triggered when the user types any character after `@`, containing that text along with the `@` character
+in the event data. For example, if a user types `@` and then adds `a` the event will be emitted with data that includes the value `@a`.
+The client can listen to this event, filter the list of users that includes the names containing the letter `a`, and then dynamically
+update the `nimble-mapping-mention-user` element based on the filter data. Subsequently, a maximum of 20 filtered options should be
+transmitted to the editor.
+
+Since the above event triggers for every key down event like adding/removing texts, move the text cursors after the `@` character which is quite an
+expensive operation to perform for every keystroke so it is advisable to `debounce` the events if you're using network requests to perform the
+filtering operations. Like, allow at most one request per second to filter the list for each second instead of for each keystrokes.
+
+_Note_: The editor will also perform the same filtering once again to ensure the filtered options are proper and update the dropdown list in the UI.
+This helps to filter the list, regardless of whether the client is loading the list dynamically by listening to the event as mentioned above
+or statically providing user details at the start via `nimble-mapping-mention`.
+
+### API
+
+To support the `@mention`, the creation of the following configuration elements and UI elements is required for rich text components:
+
+#### User mention element (Non-visible configuration element):
+
+This element serves as a container for the mapping elements and additionally functions as a configuration element. It allows the transmission of
+arbitrary values that are specific to the `@mention` users within the rich text components. Currently, this component has the below
+configuration properties/attributes for user mentions and plays the role of containing the user mapping elements.
+
+This component can be used as a feature flag to enable the support for `@mention` in the rich text editor and viewer.
+If this component is not present at the children of the rich text editor or viewer, the component will function the same without the support
+for `@mention` in any place like editor, viewer, markdown parser and serializer and the toolbar button to add **"@"** in the editor
+will be hidden.
+
+_Component Name_
+
+-   `nimble-rich-text-mention-users`
+
+_Content_
+
+-   Zero or more `nimble-mapping-mention-user` elements
+
+_Props/Attrs_
+
+-   `pattern` - is a string attribute that gets the regex pattern to match the `mention-url` in the mapping element. Exactly similar to
+    HTML attribute: [pattern](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern).
+-   `mentioned-users` - is a read-only property that returns an array of strings representing the mentioned user URLs (`mention-url` property value of
```

Comment:
Yes, all urls in the `mentioned-hrefs` are in the current state of the editor and not dependent on the mapping elements. These values will be pushed from the current element populated in the editor (DOM) and the URLs will be fetched from the `mention-href` value of the rendered view element `nimble-rich-text-mention-users-view`.

The list is maintained in `markdown-serializer` which we thought is the right place to get the mention node details. The markdown serialization is happening only in the editor as of now and not in the viewer. So the `mentioned-hrefs` will not be of use to the viewer and also we don't see any use case. However, if it is required to stay consistent in both the components we can get `mentioned-hrefs` for the viewer also.

Reply:
> So the mentioned-hrefs will not be of use to the viewer and also we don't see any use case.

What? The spec itself says under rich text viewer that users should use `mentioned-hrefs`:

> The client application should either parse the markdown string and get the user Href that match the markdown string of @mention to identify what are all the mentioned users or utilize the `mentioned-hrefs` read-only property in `nimble-rich-text-mention-users` that stores the Href of the mentioned users in an array while creating a comment or adding a description in the editor to identify the mentioned users in the markdown string.

And that seems perfectly reasonable. The client calls setMarkdown. Sees the result of changes in mentioned-hrefs and then creates the needed mapping elements.

Am I misunderstanding? Why is this not a use case in the viewer?

> The list is maintained in markdown-serializer 

Is maintaining this list an expensive operation? I would expect the property to always be accurate. If it is an accessor, is it an expensive operation that is as expensive as full serialization? If it is expensive then should it be a method instead of a property?

Reply:
Sure, since this is also an expensive operation then better not to give an option to data binding the accessor. So I have changed it to `getMentionedHrefs()` methods.